### PR TITLE
Implement P2210R2 Superior String Splitting

### PR DESCRIPTION
**Pull request by miscco**
This implements the first part of P2210R2  renaming `split_view` to `lazy_split_view`

Works towards #1980.

### CODE REVIEW

File: stl/inc/ranges [hunk 1]
```diff
@@ -3281,12 +3281,21 @@ namespace ranges {
...
-            _NODISCARD constexpr iterator_t<_BaseTy>& _Get_current() noexcept {
+            _NODISCARD constexpr iterator_t<_BaseTy> _Get_current() && noexcept {
+                if constexpr (forward_range<_BaseTy>) {
+                    return _STD move(this->_Current);
+                } else {
+                    return _STD move(*_Parent->_Current);
+                }
+            }
+
+            _NODISCARD constexpr iterator_t<_BaseTy>& _Get_current() & noexcept {
```
Comment (StephanTLavavej) [inline stl/inc/ranges:3298]: The tests are failing with:
```
In file included from C:\a\1\s\tests\std\tests\P0896R4_ranges_alg_mismatch\test.cpp:8:
D:\build\out\inc\ranges(3306,61): error: cannot overload a member function without a ref-qualifier with a member function with ref-qualifier '&'
            _NODISCARD constexpr const iterator_t<_BaseTy>& _Get_current() const noexcept {
                                                            ^
D:\build\out\inc\ranges(3298,55): note: previous declaration is here
            _NODISCARD constexpr iterator_t<_BaseTy>& _Get_current() & noexcept {
                                                      ^
D:\build\out\inc\ranges(3306,61): error: cannot overload a member function without a ref-qualifier with a member function with ref-qualifier '&&'
            _NODISCARD constexpr const iterator_t<_BaseTy>& _Get_current() const noexcept {
                                                            ^
D:\build\out\inc\ranges(3290,54): note: previous declaration is here
            _NODISCARD constexpr iterator_t<_BaseTy> _Get_current() && noexcept {
                                                     ^
2 errors generated.
```

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/ranges (hunk 1 above)
Comment (miscco) [reply stl/inc/ranges:3298]: Oh I will just add this tiny little overload, what could possibly go wrong?

Comment (miscco) [reply stl/inc/ranges]: Fixed and simplified

File: tests/std/tests/P0896R4_ranges_range_machinery/test.cpp [hunk 2]
```diff
@@ -109,7 +109,7 @@ STATIC_ASSERT(test_cpo(ranges::views::join));
 STATIC_ASSERT(test_cpo(ranges::views::keys));
 STATIC_ASSERT(test_cpo(ranges::views::reverse));
 STATIC_ASSERT(test_cpo(ranges::views::single));
-STATIC_ASSERT(test_cpo(ranges::views::split));
```
Comment (AdamBucior) [inline tests/std/tests/P0896R4_ranges_range_machinery/test.cpp:112]: This should be restored.

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 3]
```diff
@@ -13,12 +14,14 @@
 using namespace std;
 
 template <class Rng, class Delimiter>
-concept CanViewSplit = requires(Rng&& r, Delimiter&& d) {
+concept CanViewLazySplit = requires(Rng&& r, Delimiter&& d) {
```
Comment (AdamBucior) [inline tests/std/tests/P0896R4_views_split/test.cpp:17]: This should be `CanViewSplit`

**Review by AdamBucior — APPROVED** [review]

File: stl/inc/ranges [hunk 4]
```diff
@@ -1395,7 +1395,8 @@ namespace ranges {
                 return _RANGES iter_swap(_Left._Current, _Right._Current);
             }
 
-            _NODISCARD constexpr bool _Equal(const sentinel_t<_Vw>& _Last) const {
+            _NODISCARD constexpr bool _Equal(const sentinel_t<_Vw>& _Last) const
```
Comment (miscco) [inline stl/inc/ranges:1398]: Note that is an unrelated change that I found while being in the vicinity

File: stl/inc/ranges [hunk 5]
```diff
@@ -3589,9 +3608,214 @@ namespace ranges {
...
+            }
+
+            // clang-format off
+            template <class _Delim>
+                requires is_lvalue_reference_v<_Delim> || move_constructible<_Delim>
+            _NODISCARD constexpr auto operator()(_Delim&& _Delimiter) const
+                noexcept(is_lvalue_reference_v<_Delim> || is_nothrow_move_constructible_v<_Delim>) {
+                // clang-format on
+                return _Partial<_Delim>{._Delimiter = _STD forward<_Delim>(_Delimiter)};
+            }
```
Comment (timsong-cpp) [inline stl/inc/ranges:3658]: This, and the `split` version in unchanged text below, are inconsistent with WG21-P2281. Some of the examples in the discussion section of that paper may be of interest.

File: stl/inc/ranges [hunk 6]
```diff
@@ -3589,9 +3608,214 @@ namespace ranges {
...
+            struct _Partial : _Pipe::_Base<_Partial<_Delim>> {
+                /* [[no_unique_address]] */ _Delim _Delimiter;
+
+                template <viewable_range _Rng>
+                _NODISCARD constexpr auto operator()(_Rng&& _Range) const& noexcept(
+                    noexcept(lazy_split_view{_STD forward<_Rng>(_Range), _Delimiter})) requires requires {
+                    lazy_split_view{static_cast<_Rng&&>(_Range), _Delimiter};
+                }
+                { return lazy_split_view{_STD forward<_Rng>(_Range), _Delimiter}; }
+
```
Comment (timsong-cpp) [inline stl/inc/ranges:3631]: This will need a `const&& = delete;` overload to approximate perfect forwarding call wrapper behavior.

**Review by timsong-cpp — CHANGES_REQUESTED** [review]

File: stl/inc/ranges [hunk 7]
```diff
@@ -3267,8 +3267,8 @@ namespace ranges {
     template <input_range _Vw, forward_range _Pat>
         requires (view<_Vw> && view<_Pat>
             && indirectly_comparable<iterator_t<_Vw>, iterator_t<_Pat>, _RANGES equal_to>
-            && (forward_range<_Vw> || _Tiny_range<_Pat>)
-    class lazy_split_view;
+            && (forward_range<_Vw> || _Tiny_range<_Pat>))
+    class split_view;
```
Comment (AdamBucior) [inline stl/inc/ranges]: This seems like a mistake made while resolving merge conflicts. Ditto below

Comment (miscco) [reply stl/inc/ranges]: Thanks a lot!

**Comment by StephanTLavavej** [comment]
@miscco - Are you still working on resolving @timsong-cpp's comments, or were they resolved in the recent commits? Just wondering whether this should move to Initial Review again. :smiley_cat:

**Comment by miscco** [comment]
> @miscco - Are you still working on resolving @timsong-cpp's comments, or were they resolved in the recent commits? Just wondering whether this should move to Initial Review again. 😺

Actually this is something that I would want to get @CaseyCarters opinion on. As far as I remember his analysis was that we do not need to do anything related  to P2281 and if we change something here I would want to be sure we change it consistently everywhere

File: tests/std/tests/P0896R4_views_lazy_split/test.cpp [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+            Read empty{span<const char, 0>{}};
+            test_one(move_if_needed(empty), "is"sv, views::empty<string_view>);
+        }
+
+        { // LWG-3478, trailing empty range
+            Read read{span{trailing_empty}};
+            test_one(move_if_needed(read), ' ', expected_trailing);
+        }
+
+        if constexpr (ranges::forward_range<Read>) { // LWG-3505, mutlichar pattern
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_views_lazy_split/test.cpp:312]:
```suggestion
        if constexpr (ranges::forward_range<Read>) { // LWG-3505, multichar pattern
```

File: stl/inc/ranges (hunk 6 above)
Comment (CaseyCarter) [reply stl/inc/ranges:3668]: `_Delim` is a program-defined type; how can we conformingly fail to implement all four overloads of `operator()`? (Ditto for `lazy_split`'s range adaptor closure.)

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 9]
```diff
@@ -292,13 +253,23 @@ struct instantiator {
...
             Read empty{span<const char, 0>{}};
             test_one(move_if_needed(empty), "is"sv, views::empty<string_view>);
         }
+
+        { // LWG-3478, trailing empty range
+            Read read{span{trailing_empty}};
+            test_one(move_if_needed(read), ' ', expected_trailing);
+        }
+
+        { // LWG-3505, mutlichar pattern
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_views_split/test.cpp:269]: I congratulate you on your consistent spelling of "mutlichar" =)
```suggestion
        { // LWG-3505, multichar pattern
```

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/ranges (hunk 6 above)
Comment (timsong-cpp) [reply stl/inc/ranges:3668]: Conformingly, you'll need 8. See https://reviews.llvm.org/D107098#2915589 and https://reviews.llvm.org/D107199.

File: stl/inc/ranges [hunk 10]
```diff
@@ -3626,9 +3645,214 @@ namespace ranges {
...
+            requires copy_constructible<_Vw> {
+            return _Range;
+        }
+        _NODISCARD constexpr _Vw base() && noexcept(is_nothrow_move_constructible_v<_Vw>) /* strengthened */ {
+            return _STD move(_Range);
+        }
+
+        _NODISCARD constexpr auto begin() {
+            auto _First = _RANGES begin(_Range);
+            return _Iterator{*this, _First, _Find_next(_First)};
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/ranges:3841]: Urgh, this is really not pretty. We need to cache a subrange whereas our `_Cached_position` only works with an iterator.
I will cook something up, most likely a fake optional that memoizes `_Next`

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 11]
```diff
@@ -335,6 +295,26 @@ constexpr bool instantiation_test() {
     instantiator::call<test_range<forward_iterator_tag, Common::yes, CanView::yes, Copyability::move_only>>();
     instantiator::call<test_range<forward_iterator_tag, Common::yes, CanView::yes, Copyability::copyable>>();
 
+    { // ensure we get something contiguous
+        for (std::string_view sv : "127..0..0..1"sv | std::views::split(".."sv)) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_split/test.cpp:299]: This test is `using namespace std;`. Occurs below.

File: stl/inc/xutility [hunk 12]
```diff
@@ -5558,6 +5571,153 @@ namespace ranges {
     };
 
     inline constexpr _Find_if_not_fn find_if_not{_Not_quite_object::_Construct_tag{}};
+
+    // clang-format off
+    // FUNCTION TEMPLATE _Equal_rev_pred
```
Comment (StephanTLavavej) [inline stl/inc/xutility:5576]: Should remove shouty banner.

File: stl/inc/xutility [hunk 13]
```diff
@@ -5558,6 +5571,153 @@ namespace ranges {
...
+        for (; _First2 != _Last2; ++_First1, (void) ++_First2) {
+            if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_First1), _STD invoke(_Proj2, *_First2))) {
+                return {false, _It1{}};
+            }
+        }
+
+        return {true, _STD move(_First1)};
+    }
+
+    // VARIABLE ranges::search
```
Comment (StephanTLavavej) [inline stl/inc/xutility:5616]: Another shouty banner to remove.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/ranges (hunk 6 above)
Comment (CaseyCarter) [reply stl/inc/ranges:3748]: I've added an [action item](https://github.com/microsoft/STL/projects/9#card-62754027) to track fixing this in a followup PR.

File: stl/inc/ranges (hunk 5 above)
Comment (CaseyCarter) [reply stl/inc/ranges:3775]: Ditto this will be fixed in a followup.

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 14]
```diff
@@ -153,63 +156,26 @@ constexpr void test_one(Base&& base, Delimiter&& delimiter, Expected&& expected)
...
-        }
-
-        if constexpr (copyable<R>) {
-            auto r2       = r;
-            const auto i2 = r2.begin();
-            if (!is_empty) {
-                assert(ranges::equal(*i2, *i));
-            }
-        }
+    STATIC_ASSERT(!CanMemberBegin<const R>);
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_views_split/test.cpp:159]: We can strengthen this to "no `const` `begin`, member or otherwise":
```suggestion
    STATIC_ASSERT(!CanBegin<const R&>);
```

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 15]
```diff
@@ -153,63 +156,26 @@ constexpr void test_one(Base&& base, Delimiter&& delimiter, Expected&& expected)
...
-    STATIC_ASSERT(ranges::common_range<const R> == should_be_const_common);
-    const auto sc = as_const(r).end();
-    if constexpr (ranges::forward_range<Base> && ranges::forward_range<const remove_cvref_t<Base>>) {
-        STATIC_ASSERT(same_as<decltype(sc), const ranges::sentinel_t<const R>>);
-        assert((as_const(r).begin() == sc) == is_empty);
-    }
-    if constexpr (!ranges::common_range<const R>) {
-        STATIC_ASSERT(same_as<ranges::sentinel_t<R>, default_sentinel_t>);
-    }
+    STATIC_ASSERT(!CanMemberEnd<const R>);
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_views_split/test.cpp:175]: Ditto strengthen:
```suggestion
    STATIC_ASSERT(!CanEnd<const R&>);
```

**Review by CaseyCarter — APPROVED** [review]

**Comment by miscco** [comment]
@CaseyCarter I would also suggest to move this to a follow up. As far as I can tell there are two things we need to do
1. Improve the perfect forwarding for the adaptors
2. Iplement _decay-copy_ for `split_view` and `lazy_split_view`

Given that those are non trivial changes (and I currently do not have that much capacity for delicate tasks) I feel it is better to extract them into a follow up.

**Comment by StephanTLavavej** [comment]
I've pushed a merge with `main`, fixing a trivial adjacent-add conflict in `yvals_core.h` (paper numbers). There was also a fairly simple conflict in `<ranges>`, where this PR added `explicit operator bool()` and @AdamBucior's #2118 also added it with `requires _Needs_operator_bool` which defaults to `true`. I kept #2118's definition, and no further changes needed to be added to this PR due to the default.

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 16]
```diff
@@ -335,6 +295,26 @@ constexpr bool instantiation_test() {
...
+            assert(!sv.empty());
+        }
+    }
+
+    if (!is_constant_evaluated()) { // test the from_chars example
+        auto ip    = "1.2.3.4"sv;
+        auto parts = ip | views::split('.') | views::transform([](span<char const> s) {
+            int i;
+            from_chars(s.data(), s.data() + s.size(), i);
+            return i;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_split/test.cpp:309]: If something went wrong (e.g. the splitting process produced empty string_views), the `from_chars` error code isn't being checked, and `int i` is uninitialized (and unmodified in the event of failure). I believe that initializing to `0` and checking the `errc` would be the most robust thing to do, and it's only a couple extra lines.

File: tests/std/tests/P0896R4_views_lazy_split/test.cpp [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+    constexpr bool should_be_const_common = ranges::forward_range<Base>
+        && ranges::forward_range<const remove_cvref_t<Base>> && ranges::common_range<const remove_cvref_t<Base>>;
+    // clang-format on
+    STATIC_ASSERT(ranges::common_range<const R> == should_be_const_common);
+    const auto sc = as_const(r).end();
+    if constexpr (ranges::forward_range<Base> && ranges::forward_range<const remove_cvref_t<Base>>) {
+        STATIC_ASSERT(same_as<decltype(sc), const ranges::sentinel_t<const R>>);
+        assert((as_const(r).begin() == sc) == is_empty);
+    }
+    if constexpr (!ranges::common_range<const R>) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_lazy_split/test.cpp:212]: Style, should have a newline here.

**Review by StephanTLavavej — COMMENTED** [review]
Thanks! I'll push changes to address my feedback (they were all localized).

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/ranges [hunk 18]
```diff
@@ -3746,9 +3765,220 @@ namespace ranges {
...
+            }
+        };
+
+        class _Sentinel {
+        private:
+            /* [[no_unique_address]] */ sentinel_t<_Vw> _Last{};
+
+            _NODISCARD constexpr bool _Equal(const _Iterator& _It) const
+                noexcept(noexcept(_Implicitly_convert_to<bool>(_It._Current == _Last))) {
+                return !_It._Trailing_empty && _It._Current == _Last;
```
[the remark this answers is not in the document]
Comment (timsong-cpp) [reply stl/inc/ranges:3906]: I think that's debatable - most of the time, `x.cur_ == y.end_` is `false` and `x.trailing_empty_` is also `false`. Testing the `bool` first is actually extra work.

**Comment by miscco** [comment]
Thanks a lot for the review and the fixes 😸

**Comment by StephanTLavavej** [comment]
I'm mirroring this to an MSVC-internal PR - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
`"Thanks", "For", "Implementing", "This", "C++20", "Defect", "Report"` :smile_cat:
